feat: LDAP configuration

Gu1nness opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- markdownlint-disable MD041 -->

## 🏷 Type of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update
- [ ] Tooling and CI changes
- [ ] Dependencies upgrade or change

## 📝 Description
<!--
Describe your changes in detail including:
  - the purpose and scope of this PR (what)
  - the impacted components (where, often match conventional commit scope)
  - explanation/algorithm if require (how)
-->
This PR updates the configuration managers for mongodb and mongos to conditionally add the ldap configuration for the service.
It also adds some testing for this configuration and some full cycle unit testing.
Finally, it updates the restart mechanism to restart only if needed.

## 📑 Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- Either the associated JIRA reference or a paragraph. -->
This is one of the final steps of LDAP integration.

## 🧪 How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
This has been unit tested and manually tested on my machine.

## ✅ Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](../blob/6/edge/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Gu1nness committed:
feat: Use config file instead of cli parameters

Gu1nness committed:
fix: extend documentation

Gu1nness committed:
fix: bump deps

Gu1nness committed:
fix: split paths + chmod

Gu1nness committed:
fix: output destination + remove useless string

Gu1nness committed:
fix: unit test

Gu1nness committed:
fix: Already bump version

Gu1nness committed:
fix: remove now useless check

Gu1nness committed:
wip: DPE-6715

Gu1nness committed:
fix: remove useless parameter

Gu1nness committed:
fix: don't merge config

Gu1nness committed:
fix: journal since we use only built config

Gu1nness committed:
fix: test

Gu1nness committed:
fix: progess

Gu1nness committed:
fix: sleep to avoid race condition

Gu1nness committed:
Merge branch '6/edge' into feat/DPE-6715-ldap-manager

Gu1nness committed:
doc: Explain why the build_parameters function is empty

Gu1nness committed:
fix: post review

Gu1nness committed:
feat: events

Gu1nness committed:
fix: unit test

Gu1nness committed:
Merge branch '6/edge' into feat/DPE-6717-mongo-db-ldap-update-configuration-to-write-in-files-instead-of-cli-mongodb-mongos

Gu1nness committed:
fix: quite a few fixes

Gu1nness committed:
fix: mongos requires configDB

Gu1nness committed:
fix: unit tests

Gu1nness committed:
fix: update charmcraft file

Gu1nness committed:
fix: change status

Gu1nness committed:
Merge branch 'feat/DPE-6717-mongo-db-ldap-update-configuration-to-write-in-files-instead-of-cli-mongodb-mongos' into feat/DPE-6791-update-config-manager-ldap

Gu1nness committed:
feat: don't use system wide certificates

Gu1nness committed:
fix: move get_ldap_connection_status

Gu1nness committed:
Merge branch 'feat/DPE-6715-ldap-manager' into feat/DPE-6716-mongo-db-ldap-create-events-for-gl-auth-k-8-s-integration

Gu1nness committed:
fix: better status

Gu1nness committed:
feat: bump snap revision

Gu1nness committed:
feat: LDAP configuration

Gu1nness committed:
fix: mock restart

Gu1nness committed:
fix: small inconsistencies

Gu1nness committed:
fix: force restart on cert remove

Gu1nness committed:
fix: use peer data

Gu1nness committed:
test: test ldap config

MiaAltieri commented on single_kernel_mongo/managers/config.py:
```diff
@@ -87,6 +87,17 @@ def set_environment(self):
 
+    def configure_and_restart(self, force: bool = False) -> None:
+        """Re-configure if needed and restart the service if needed."""
+        data = "\n".join(self.workload.read(self.file))
+        current_content = safe_load(data)
+
+        new_content = self.build_config()
+
+        if force or not self.workload.active() or new_content != current_content:
+            self.workload.write(self.file, safe_dump(new_content))
+            self.workload.restart()
```
nit readability 
```suggestion
        current_config_file = "\n".join(self.workload.read(self.file))
        current_config_file_content = safe_load(current_config_file)

        new_content = self.build_config()

        if force or not self.workload.active() or new_content != current_config_file_content:
            self.workload.write(self.file, safe_dump(new_content))
            self.workload.restart()
```

MiaAltieri commented on single_kernel_mongo/managers/config.py:
```diff
@@ -393,6 +411,36 @@ def replset_option(self) -> dict[str, Any]:
+    def ldap_parameters(self) -> dict[str, Any]:
+        if not self.state.ldap.is_ready():
+            return {}
+        if self.state.is_role(MongoDBRoles.SHARD):
```
nit add a comment as to why (explaining your expert knowledge)

MiaAltieri commented on single_kernel_mongo/managers/config.py:
```diff
@@ -430,6 +478,28 @@ def config_server_db_parameter(self) -> dict[str, Any]:
 
+    @property
+    @override
+    def ldap_parameters(self) -> dict[str, Any]:
```
can you add in the docstring why this is needed when there is already one for replica set (for future readers + maintainablitiy)

Gu1nness replied:
I don't understand the question, this is the config for mongos.
Explained in the [spec](https://docs.google.com/document/d/1J483J987gaY1znCQE79VEucwgdWO1RkWa5gRdkGpOT0/edit?tab=t.0) that:
 * config server is configured fully because it allows direct connection + it's responsible for authorization in case of sharded cluster
 * mongos is configure for authentication.
 
Usually, identity users will use a mongos to connect because they want to be able to access data on all shards, taking advantage of the mongos router. But authorization is still performed by the config-server, mongos is just a router.

Is that what you're asking?

MiaAltieri replied:
Ah - yes - sorry I had confused myself - thanks for clearing things up! Resolving

MiaAltieri commented on single_kernel_mongo/managers/mongodb_operator.py:
```diff
@@ -422,7 +422,8 @@ def on_config_changed(self) -> None:
                 self.state.app_peer_data.ldap_query_template = self.config.ldap_query_template
 
-            # TODO: Invalidate the cache and restart with those parameters if needed.
+            # This will restart only if the config was changed.
```
love it

MiaAltieri reviewed (commented):
Looks great! Nits only!

Gu1nness committed:
fix: post review

Gu1nness committed:
Merge branch 'feat/DPE-6715-ldap-manager' into feat/DPE-6716-mongo-db-ldap-create-events-for-gl-auth-k-8-s-integration

Gu1nness committed:
Update single_kernel_mongo/managers/config.py

Co-authored-by: Mia Altieri <32723809+MiaAltieri@users.noreply.github.com>

Gu1nness committed:
Merge branch 'feat/DPE-6716-mongo-db-ldap-create-events-for-gl-auth-k-8-s-integration' into feat/DPE-6791-update-config-manager-ldap

Gu1nness committed:
fix: doc

Gu1nness committed:
Merge branch 'feat/DPE-6715-ldap-manager' into feat/DPE-6716-mongo-db-ldap-create-events-for-gl-auth-k-8-s-integration

Gu1nness committed:
Merge branch 'feat/DPE-6716-mongo-db-ldap-create-events-for-gl-auth-k-8-s-integration' into feat/DPE-6791-update-config-manager-ldap

Mehdi-Bendriss commented on single_kernel_mongo/managers/config.py:
```diff
@@ -430,6 +481,28 @@ def config_server_db_parameter(self) -> dict[str, Any]:
+                        "queryUser": self.state.ldap.bind_user,
+                        "queryPassword": self.state.ldap.bind_password,
+                    },
+                    # TODO: Update this when writing config-server/mongos sync.
```
Could you explain why is `authz` missing here?

Gu1nness replied:
`authz` is used for authorisation, which is done by the config server and not the mongos.
This specific option [is not available on mongos](https://www.mongodb.com/docs/manual/reference/configuration-options/#mongodb-setting-security.ldap.authz.queryTemplate)

Mehdi-Bendriss commented on single_kernel_mongo/managers/mongodb_operator.py:
```diff
@@ -807,16 +808,14 @@ def stop_charm_services(self):
 
         If we are running as config-server, we should update both mongod and mongos environments.
         """
-        self.stop_charm_services()
-        self.config_manager.set_environment()
+        self.config_manager.configure_and_restart(force=force)
```
nice

Mehdi-Bendriss commented on single_kernel_mongo/managers/mongodb_operator.py:
```diff
@@ -807,16 +808,14 @@ def stop_charm_services(self):
         if self.state.is_role(MongoDBRoles.CONFIG_SERVER):
-            self.mongos_config_manager.set_environment()
-        self.start_charm_services()
+            self.mongos_config_manager.configure_and_restart(force=force)
```
This seems like we are restarting twice, no?

Gu1nness replied:
No, we restart the mongod and mongos if we are a config-server;
On the config-server, there is a mongod and a mongos running, the mongos running being used only by the charm itself.
We however have to restart both services when needed.

Mehdi-Bendriss reviewed (commented):
Thanks Neha! Looks good! I have a couple of questions

Gu1nness committed:
Merge branch 'feat/DPE-6715-ldap-manager' into feat/DPE-6791-update-config-manager-ldap

Gu1nness committed:
feat(DPE-6792): Share config with mongos (#33)

Instead of requiring the operator to config the two services, the config-server shares the ldapUserToDNMapping with mongos through the cluster relation.
It also shares a hash of the ca chain + ldap addresses to ensure that both services are connected with the same ldap server.
